### PR TITLE
Create a separate Python/C helper library to remove code duplication

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Parts of the code base had duplicated the GIL class and error handling code from the Framework PythonInterface code. The new library required for supporting the matplotlib-based instrument view plotter will requires these features too so it's time to make them easily accessible.

Over time this library should grow with most code moving out of PythonInterface kernel.

**Report to:** [nobody]. 

**To test:**

* Code review
* Note that the large number of changes in `Mantid_template_doxyfile` is due to whitespace fixes.

*This does not require release notes* because **it is an internal change.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
The failing system tests are currently being fixed in #23579.

File: Framework/PythonInterface/core/src/GlobalInterpreterLock.cpp [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
 //------------------------------------------------------------------------------
 // GlobalInterpreterLock Static helpers
 //------------------------------------------------------------------------------
 
+/**
+ * Check if the current thread has the lock
+ * @return True if the current thread holds the GIL, false otherwise
+ */
+bool GlobalInterpreterLock::locked() {
+#if PY_VERSION_HEX < 0x03000000
```
Comment (peterfpeterson) [inline Framework/PythonInterface/core/src/GlobalInterpreterLock.cpp]: shouldn't this use `IS_PY3K`?

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Overall good reduction in duplication of code. One minor change.

File: Framework/PythonInterface/core/src/GlobalInterpreterLock.cpp (hunk 1 above)
Comment (martyngigg) [reply Framework/PythonInterface/core/src/GlobalInterpreterLock.cpp]: Nice catch. Yes it should.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**